resolwe/flow/migrations/0037_migrate_m2o.py

```diff
@@ -0,0 +1,227 @@
+from django.core.exceptions import ValidationError
+from django.db import migrations, models
+
+from django.utils.timezone import now
+
+from resolwe.flow.utils import rewire_inputs
+
+
+def copy_permissions(original, duplicate, model, apps):
+    """Copy permissions from original object to it's duplicate."""
+    UserObjectPermission = apps.get_model('guardian', 'UserObjectPermission')
+
+    kwargs = {
+        'permission__content_type__model': model,
+        'object_pk': original.id,
+    }
+    for perm in UserObjectPermission.objects.filter(**kwargs):
+        perm.id = None
+        perm.object_pk = duplicate.id
+        perm.content_object = duplicate
+        perm.save(force_insert=True)
+
+
+def duplicate_entity(original, apps):
+    """Duplicate entity."""
+    Entity = apps.get_model('flow', 'Entity')
+
+    duplicate = Entity.objects.get(id=original.id)
+    duplicate.pk = None
+    duplicate.slug = None
+    duplicate.collection = None
+    duplicate.name = 'Copy of {}'.format(original.name)
+    duplicate.duplicated = now()
+
+    duplicate.save(force_insert=True)
+
+    copy_permissions(original, duplicate, model='entity', apps=apps)
+
+    # Override fields that are automatically set on create.
+    duplicate.created = original.created
+    duplicate.save()
+
+    # Duplicate entity's data objects.
+    bundle = [{'original': obj, 'copy': duplicate_data(obj, apps)} for obj in original.data.all()]
+    bundle = rewire_inputs(bundle)
+    duplicated_data = [item['copy'] for item in bundle]
+
+    duplicate.data.add(*duplicated_data)
+
+    return duplicate
+
+
+def duplicate_data(original, apps):
+    """Duplicate (make a copy)."""
+    DataDependency = apps.get_model('flow', 'DataDependency')
+    Data = apps.get_model('flow', 'Data')
+
+    # TODO
+    # if original.status not in ['OK', 'ER']:
+    #     raise ValidationError('Data object must have done or error status to be duplicated')
+
+    duplicate = Data.objects.get(id=original.id)
+    duplicate.pk = None
+    duplicate.slug = None
+    duplicate.entity2 = None
+    duplicate.collection = None
+    duplicate.name = 'Copy of {}'.format(original.name)
+    duplicate.duplicated = now()
+
+    duplicate.save(force_insert=True)  # pylint: disable=protected-access
+
+    copy_permissions(original, duplicate, model='data', apps=apps)
+
+    # Override fields that are automatically set on create.
+    duplicate.created = original.created
+    duplicate.save()  # pylint: disable=protected-access
+
+    if original.location:
+        original.location.data.add(duplicate)  # pylint: disable=no-member
+
+    duplicate.storages.set(original.storages.all())  # pylint: disable=no-member
+
+    for migration in original.migration_history.order_by('created'):  # pylint: disable=no-member
+        migration.pk = None
+        migration.data = duplicate
+        migration.save(force_insert=True)
+
+    # Inherit existing child dependencies.
+    DataDependency.objects.bulk_create([
```

Comment:
If whole entity/collection is duplicated, Data objects should have dependencies between themselves, not with original objects.

Reply:
True. But this is solved in by function `rewire_inputs` in `duplicate_entity`.